editoast/src/schema/railjson.rs

```diff
@@ -0,0 +1,74 @@
+use thiserror::Error;
+
+use derivative::Derivative;
+use diesel::PgConnection;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    api_error::ApiError,
+    infra::{Infra, RAILJSON_VERSION},
+};
+
+use super::{
+    BufferStop, Catenary, Detector, OperationalPoint, Route, Signal, SpeedSection, Switch,
+    SwitchType, TrackSection, TrackSectionLink,
+};
+
+#[derive(Deserialize, Derivative, Serialize, Clone, Debug)]
+#[derivative(Default)]
+#[serde(deny_unknown_fields)]
+pub struct RailJson {
+    pub version: String,
+    pub operational_points: Vec<OperationalPoint>,
+    pub routes: Vec<Route>,
+    pub switch_types: Vec<SwitchType>,
+    pub switches: Vec<Switch>,
+    pub track_section_links: Vec<TrackSectionLink>,
+    pub track_sections: Vec<TrackSection>,
+    pub speed_sections: Vec<SpeedSection>,
+    pub catenaries: Vec<Catenary>,
+    pub signals: Vec<Signal>,
+    pub buffer_stops: Vec<BufferStop>,
+    pub detectors: Vec<Detector>,
+}
+
+#[derive(Debug, Error)]
+pub enum RailjsonError {
+    #[error("Wrong railjson version '{0}'. Should be {}", RAILJSON_VERSION)]
+    RailjsonVersion(String),
+}
+
+impl RailJson {
+    pub fn persist_batch(
+        &self,
+        infrastructure_name: String,
+        conn: &PgConnection,
+    ) -> Result<Infra, Box<dyn ApiError>> {
+        let result = conn
+            .build_transaction()
+            .read_write()
+            .run::<Infra, Box<dyn ApiError>, _>(|| -> Result<Infra, Box<dyn ApiError>> {
+                let infra =
+                    Infra::create_with_railjson_version(&infrastructure_name, &self.version, conn)
+                        .unwrap();
+
+                // IMPROVE could implement a generic persist method for both elements
+                BufferStop::persist(&self.buffer_stops, infra.id, conn)?;
+                Catenary::persist(&self.catenaries, infra.id, conn)?;
+                Detector::persist(&self.detectors, infra.id, conn)?;
+                OperationalPoint::persist(&self.operational_points, infra.id, conn)?;
+                Route::persist(&self.routes, infra.id, conn)?;
+                Signal::persist(&self.signals, infra.id, conn)?;
+                Switch::persist(&self.switches, infra.id, conn)?;
+                SwitchType::persist(&self.switch_types, infra.id, conn)?;
+                TrackSectionLink::persist(&self.track_section_links, infra.id, conn)?;
+                TrackSection::persist(&self.track_sections, infra.id, conn)?;
+                Ok(infra)
+            });
+
+        match result {
+            Ok(infra) => Ok(infra),
+            Err(err) => Err(err),
+        }
```

Comment:
Bad pattern.

```suggestion
       result
```

Actually you can return the result of the closure without variable

Reply:
ok